stdlib/src/collections/inline_array.mojo

```diff
@@ -326,7 +326,7 @@ struct InlineArray[
     @always_inline
     fn __contains__[
         T: EqualityComparableCollectionElement, //
-    ](self, value: T) -> Bool:
+    ](self: InlineArray[T, size], value: T) -> Bool:
```

Comment:
This is a great way to get rid of the `constrained` check later, which shifts the error message from the elaborator (which produces terrible stack traces) up to parser time.  Could we use this in more places?

Reply:
@soraros curious your view on this, I'm a fan :-)

Reply:
Indeed we can, and I did some in #3210 in the past.

However, we can run into problems:
1. Heterogeneous container, example: [`Tuple.__getitem__`](https://github.com/modularml/mojo/blob/c687893d202d59f1556d1cb370a0e9f37d1984cf/stdlib/src/builtin/tuple.mojo#L195-L227)
2. Trait bound mismatch, example: `InlineList[T]` takes `T: CollectionElementNew` but nobody got around to add a `EqualityComparableCollectionElementNew` (only have `EqualityComparableCollectionElement` which is slightly different) to put on its `count` function. It is only temporary since we are in the process of moving towards `CollectionElementNew`. All that is to say that I really want anonymous trait bound (`T: A & B`).
3. We are missing "half" of the support for conditional conformance: we don't have conditional extension. This one covers the majority of cases where `constrained` is needed in the stdlib.
```mojo
trait Stringable:
  fn __str__(self) -> String: ...

struct List[T]:  # List itself doesn't conform to Stringable
  fn __str__[T: Stringable](self: List[T]) -> String: ...
```

I think there aren't many cases (if any at all module 2. above) where we can use CC to remove `constrained` or `_type_is_eq`.